pmd-java/src/test/java/net/sourceforge/pmd/coverage/PMDCoverageTest.java

```diff
@@ -4,93 +4,84 @@
     }
 
     /**
      * Run the PMD command line tool, i.e. call PMD.main().
      *
      * @param commandLine
      */
-    private void runPmd(String commandLine) {
+    private void runPmd(String inputPath, Consumer<PMDConfiguration> configure) {
         StringBuilder report = new StringBuilder("missing report");
 
         try {
             Path f = Files.createTempFile(tempFolder, PMDCoverageTest.class.getSimpleName(), null);
-            String[] args = ArrayUtils.addAll(
-                commandLine.split("\\s"),
-                "-reportfile",
-                f.toAbsolutePath().toString(),
-                "-threads",
-                String.valueOf(Runtime.getRuntime().availableProcessors())
-            );
-
-            System.err.println("Running PMD with: " + Arrays.toString(args));
+
             String output = SystemLambda.tapSystemOut(() -> {
                 String errorOutput = SystemLambda.tapSystemErr(() -> {
-                    PMD.runPmd(args);
+                    PMDConfiguration conf = new PMDConfiguration();
+                    conf.addInputPath(Paths.get(inputPath));
+                    conf.setReportFile(f);
+                    conf.addRuleSet("rulesets/internal/all-java.xml");
+                    conf.setThreads(Runtime.getRuntime().availableProcessors());
+                    configure.accept(conf);
+
+                    try (PmdAnalysis pmd = PmdAnalysis.create(conf)) {
+                        pmd.performAnalysis();
+                    }
 
                     report.setLength(0);
                     report.append(IOUtil.readFileToString(f.toFile(), StandardCharsets.UTF_8));
                 });
-                assertEquals(0, StringUtils.countMatches(errorOutput, "Exception applying rule"), "No exceptions expected");
-                assertFalse(errorOutput.contains("Ruleset not found"), "Wrong configuration? Ruleset not found");
-                assertEquals(0, StringUtils.countMatches(errorOutput, "Use of deprecated attribute"), "No usage of deprecated XPath attributes expected");
+                assertThat(errorOutput, not(containsString("Exception applying rule")));
+                assertThat(errorOutput, not(containsString("Ruleset not found")));
+                assertThat(errorOutput, not(containsString("Use of deprecated attribute")));
             });
 
-            assertEquals(0, output.length(), "Nothing should be output to stdout");
-
-            assertEquals(0, StringUtils.countMatches(report, "Error while processing"), "No processing errors expected");
-
+            assertThat(output, is(emptyString()));
+            assertThat(output, not(containsString("Error while processing")));
             // we might have explicit examples of parsing errors, so these are maybe false positives
-            assertEquals(0, StringUtils.countMatches(report, "Error while parsing"), "No parsing error expected");
+            assertThat(output, not(containsString("Error while parsing")));
```

Comment:
should be `report` instead of output

Reply:
fixed with d5a44da575b32cb0ddd9f0de71e10d9cf69e9946

Reply:
Thanks Andreas!